### PR TITLE
Add a Retriever module

### LINKED ISSUE #16: Add a retriever module for DSPy
The natural language search underlying the retriever should do the following process:
```py
def search(query: text):
  similar_nodes = search_similar_nodes(query, k=10) # user can use either top-k or filter by threshold
  similar_edges = search_similar_edges(query, k=10) # user can use either top-k or filter by threshold

  resultant_subgraph = Graph()

  resultant_subgraph.add(similar_nodes)
  for node in similar_nodes:
    resultant_subgraph.add(all_connected_nodes(node))
  for edge in similar_edges:
    resultant_subgraph.add(all_connected_nodes(edge))

  return resultant_subgraph
```

### PR DESCRIPTION
**Pull request by anubhuti24**
closes #16

### CODE REVIEW

File: scripts/retriever.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+import dspy
+from dsp.utils import dotdict
+from dotenv import load_dotenv
+
+load_dotenv()
+
+sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))
+
+from personal_graph.graph import Graph  # noqa: E402
```
Comment (sutyum) [inline scripts/retriever.py:22]: why the noqa is required?

File: scripts/retriever.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        embedding_model: str = "text-embedding-ada-002",
+    ):
+        self.k = k
+        self.embedder = Embedder(embedding_provider, embedding_model)
+        self.db_url = db_url
+        self.auth_token = auth_token
+        self.graph = Graph(db_url, auth_token)
+
+    def _retrieve_passages(self, queries: List[str]) -> List[dotdict]:
+        passages = []
```
Comment (sutyum) [inline scripts/retriever.py:65]: check and act on empty list query

File: scripts/retriever.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        for query in queries:
+            kg = self.graph.search_query(query)
+            passages.extend(kg.nodes)
+        return passages
+
+    def retrieve(self, queries: Union[str, List[str]]) -> List[dotdict]:
```
Comment (sutyum) [inline scripts/retriever.py:71]: a module has to have a forward method

**Review by sutyum — CHANGES_REQUESTED** [review]

**Comment by sutyum** [comment]
- [ ] Add an example for the retriever
- [ ] Fix the implementation of natural language search used by the retriever

File: scripts/retriever.py (hunk 3 above)
Comment (anubhuti24) [reply scripts/retriever.py]: Modified it

File: personal_graph/database.py [hunk 4]
```diff
@@ -760,3 +743,38 @@ def _identical_nodes(cursor, connection):
         return similar_rows
 
     return _identical_nodes
+
+
+def all_connected_nodes(
+    node_or_edge: Tuple[Any], node_or_edge_type: str
```
Comment (sutyum) [inline personal_graph/database.py:749]: probably need a Union type here

File: personal_graph/models.py [hunk 5]
```diff
@@ -21,18 +25,30 @@ class Edge(BaseModel):
...
 class EdgeInput(BaseModel):
     source: Node = Field(..., description="Source node from which the edge originates.")
     target: Node = Field(..., description="Target node to which the edge points.")
-    label: str = Field(..., description="Label associated with the edge.")
+    label: str = Field(
+        ...,
+        min_length=2,
+        description="Most related and unique name associated with the edge.",
+    )
     attribute: Union[str, Dict[str, str]] = Field(
```
Comment (sutyum) [inline personal_graph/models.py:48]: should be 'attributes'

**Review by sutyum — CHANGES_REQUESTED** [review]

File: personal_graph/database.py (hunk 4 above)
Comment (anubhuti24) [reply personal_graph/database.py]: Updated it to use Union type

File: personal_graph/models.py (hunk 5 above)
Comment (anubhuti24) [reply personal_graph/models.py]: Modified

File: scripts/retriever.py
Comment (sutyum) [inline scripts/retriever.py:1]: this is not meant to be a script, should go in personal_graph dir

File: scripts/retriever.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))
+from personal_graph.graph import Graph
+from personal_graph.models import Node
+from dotenv import load_dotenv
+
+load_dotenv()
+
+
+class Retriever(dspy.Retrieve):
```
Comment (sutyum) [inline scripts/retriever.py:14]: Better to name it in a manner similar to other RMs in DSPy: PersonalRM

File: scripts/retriever.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+            return passages
+
+        for query in queries:
+            kg = self.graph.search_query(query)
+            self.graph.visualize_graph(kg)
+            passages.extend(kg.nodes)
+        return passages
+
+    def forward(
+        self, query_or_queries: Union[str, List[str]], k: Optional[int] = None, **kwargs
```
Comment (sutyum) [inline scripts/retriever.py:40]: add a TODO comment to use the value of k in the future

**Review by sutyum — CHANGES_REQUESTED** [review]

File: examples/low_level_api.py [hunk 8]
```diff
@@ -1,8 +1,11 @@
 #! /usr/bin/python
 import os
+import sys
 import logging
 import argparse
 from dotenv import load_dotenv
+
+sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))
```
Comment (sutyum) [inline examples/low_level_api.py:8]: why is this needed?

**Review by sutyum — CHANGES_REQUESTED** [review]

**Review by sutyum — APPROVED** [review]

File: examples/low_level_api.py (hunk 8 above)
Comment (sutyum) [reply examples/low_level_api.py]: not needed, ill explain

**Pull request merged**